src/asm/queue_add.c

```diff
@@ -0,0 +1,113 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   queue_add.c                                        :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: asarandi <asarandi@student.42.us.org>      +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2018/02/11 20:53:20 by asarandi          #+#    #+#             */
+/*   Updated: 2018/02/28 02:03:52 by asarandi         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "asm.h"
+
+void	add_label_to_queue(t_asm *a, char *label)
+{
+	t_instruction	*ptr;
+	t_instruction	*tail;
+
+	ptr = ft_memalloc(sizeof(t_instruction));
+	ptr->type = T_LAB;
+	ptr->label = ft_strdup(label);
+	ptr->label[ft_strlen(ptr->label) - 1] = 0;
+	ptr->original = a->split[a->lines - 1];
+	ptr->address = a->prog_size;
+	if (a->instruction == NULL)
+		a->instruction = ptr;
+	else
+	{
+		tail = a->instruction;
+		while (tail->next != NULL)
+			tail = tail->next;
+		tail->next = ptr;
+	}
+	return ;
+}
+
+t_in	*queue_prepare(t_asm *a, char *name, char *oper)
+{
+	t_op			instr;
+	t_instruction	*ptr;
+
+	instr = find_instruction(name);
+	ptr = ft_memalloc(sizeof(t_instruction));
+	ptr->type = T_REG;
+	ptr->original = a->split[a->lines - 1];
+	ptr->opcode = instr.opcode;
+	ptr->has_acb = instr.arg_descript;
+	ptr->address = a->prog_size;
+	ptr->i_name = ft_strdup(name);
+	ptr->i_oper = ft_strdup(oper);
+	ptr->operands = ft_strsplit(oper, SEPARATOR_CHAR);
+	if (count_char_array(ptr->operands) != instr.arg_count)
+		ft_printf("{red}ERROR:{eoc} wrong number of operands\n");
+	ptr->arg_count = instr.arg_count;
+	ptr->short_direct = instr.short_direct;
+	ptr->size = 1 + instr.arg_descript;
+	return (ptr);
+}
+
+void	add_to_queue(t_asm *a, t_in *ptr)
+{
+	t_instruction	*tail;
+
+	if (a->instruction == NULL)
```

Comment:
This function is duplicated on line 26 of this same file.

I'd also say something about preferring the use of t_array for the instructions instead of a linked list; is there a prior discussion on the tradeoffs of this?

Reply:
function is not duplicated, one adds labels to queue, the other adds instructions;
the check is to see if the beginning of linked list is null, i.e. first instruction/label to be added to queue; don't know anything about t_array, is that something custom you built?

Reply:
You can pass a pointer / boolean / etc to the function to say what you want it to add.

Reply:
A generic function that adds to the proper instructions/labels list solves this problem.